packages/vm/src/cache.rs

```diff
@@ -96,13 +96,15 @@ where
         id: &[u8],
         deps: Extern<S, A, Q>,
         gas_limit: u64,
+        allow_db_write: bool,
     ) -> Result<Instance<S, A, Q>, Error> {
         let hash = WasmHash::generate(&id);
 
         // pop from lru cache if present
         if let Some(cache) = &mut self.instances {
             if let Some(cached_instance) = cache.pop(&hash) {
                 self.stats.hits_instance += 1;
+                // Cannot override db writable flag here
                 return Ok(Instance::from_wasmer(cached_instance, deps, gas_limit));
```

Comment:
Ah, here we're running into trouble. An instance is either created with all imports or with read-only imports, meaning handle instances and query instances cannot share cache entries.

To prevent that, we could hash the writable flag into the cache key.

Reply:
Ah, I saw below. `imports` are sent to the `wasmer::Instance` upon creation. But `deps` and `gas_limit` can be modified when we wrap the wasmer `Instance` into a cosmwasm `Instance`.

I like the idea of using the writable flag in the cache key. Since this is an in-memory cache (no disk), we can just use a tuple `(hash, writable_flag)`, which implements `Hash` if both elements do. (And is clearer than concatenating them somehow)